astromodels/functions/functions.py

```diff
@@ -1367,3 +1371,313 @@ def _set_units(self, x_unit, y_unit):
+                fix : yes
+        """
+    
+    __metaclass__ = FunctionMeta
+    
+    def _setup(self):
+        
+        astroDir = astromodels.__file__
+        astroDir = astroDir.split('__init__.py')[0]
+        tablepath = astroDir+'functions/gammamc_dif.dat'
+        self._data = np.loadtxt(tablepath)
+        
+        """
+            Mapping between the channel codes and the rows in the gammamc file
+            
+            1 : 8, # ee
+            2 : 6, # mumu
+            3 : 3, # tautau
+            4 : 1, # bb
+            5 : 2, # tt
+            6 : 7, # gg
+            7 : 4, # ww
+            8 : 5, # zz
+            9 : 0, # cc
+            10 : 10, # uu
+            11 : 11, # dd
+            12 : 9, # ss
+        """
+        
+        channel_index_mapping = {
+            1 : 8, # ee
+            2 : 6, # mumu
+            3 : 3, # tautau
+            4 : 1, # bb
+            5 : 2, # tt
+            6 : 7, # gg
+            7 : 4, # ww
+            8 : 5, # zz
+            9 : 0, # cc
+            10 : 10, # uu
+            11 : 11, # dd
+            12 : 9, # ss
+        }
+        
+        # Number of decades in x = log10(E/M)
+        ndec = 10.0
+        xedge = np.linspace(0,1.0,251)
+        self._x = 0.5*(xedge[1:]+xedge[:-1])*ndec - ndec
+
+        ichan = channel_index_mapping[int(self.channel.value)]
+
+        # These are the mass points
+        self._mass = np.array([2.0,4.0,6.0,8.0,10.0,
+                       25.0,50.0,80.3,91.2,100.0,
+                       150.0,176.0,200.0,250.0,350.0,500.0,750.0,
+                       1000.0,1500.0,2000.0,3000.0,5000.0,7000.0,1E4])
+        self._dn = self._data.reshape((12,24,250))
+            
+        self._dn_interp = RegularGridInterpolator([self._mass,self._x],
+                                                    self._dn[ichan,:,:],
+                                                    bounds_error=False,
+                                                    fill_value=None)
+
+        if self.mass.value > 10000:
+            print "Warning: DMFitFunction only appropriate for masses <= 10 TeV"
+            print "To model DM from 2 GeV < mass < 1 PeV use DMSpectra"
+
+    def _set_units(self, x_unit, y_unit):
+        
+        self.mass.unit = u.GeV
+        self.channel.unit = astropy_units.dimensionless_unscaled
+        self.sigmav.unit = u.cm**3 / u.s
+        self.J.unit = u.GeV**2 / u.cm**5
+    
+    def print_channel_mapping(self):
+    
+        channel_mapping = {
+        1 : 'ee',
+        2 : 'mumu',
+        3 : 'tautau',
+        4 : 'bb',
+        5 : 'tt',
+        6 : 'gg',
+        7 : 'ww',
+        8 : 'zz',
+        9 : 'cc',
+        10 : 'uu',
+        11 : 'dd',
+        12 : 'ss',
+        }
+    
+        print channel_mapping
+    
+        return channel_mapping
+    
+    # noinspection PyPep8Naming
+    def evaluate(self, x, mass,channel,sigmav,J):
+        
+        keVtoMeV = 1./1000.
```

Comment:
The call to evaluate() should work also when x has units on its own. Doing a conversion like this will not work well in that case. I am going to fix this after merging.